File: ros2/camera_control/camera_control/camera/lucid_frame.py
from typing import Optional, Tuple

import cv2
import numpy as np

from .calibration import distort_pixel_coords, invert_extrinsic_matrix
from .rgbd_frame import RgbdFrame

# General min and max possible depths
DEPTH_MIN_MM = 500
DEPTH_MAX_MM = 10000


class LucidFrame(RgbdFrame):
    color_frame: np.ndarray
    depth_frame: np.ndarray
    timestamp_millis: float

    def __init__(
        self,
        color_frame: np.ndarray,
        depth_frame_xyz: np.ndarray,
        timestamp_millis: float,
        color_camera_intrinsic_matrix: np.ndarray,
        color_camera_distortion_coeffs: np.ndarray,
        depth_camera_intrinsic_matrix: np.ndarray,
        depth_camera_distortion_coeffs: np.ndarray,
        xyz_to_color_camera_extrinsic_matrix: np.ndarray,
        xyz_to_depth_camera_extrinsic_matrix: np.ndarray,
    ):
        """
        Args:
            color_frame (np.ndarray): The color frame in RGB8 format.
            depth_frame_xyz (np.ndarray): The depth frame in Coord3D_ABC16 format.
            timestamp_millis (float): The timestamp of the frame, in milliseconds since the device was started.
            color_camera_intrinsic_matrix (np.ndarray): Intrinsic matrix of the color camera.
            color_camera_distortion_coeffs (np.ndarray): Distortion coefficients of the color camera.
            depth_camera_intrinsic_matrix (np.ndarray): Intrinsic matrix of the depth camera.
            depth_camera_distortion_coeffs (np.ndarray): Distortion coefficients of the depth camera.
            xyz_to_color_camera_extrinsic_matrix (np.ndarray): Extrinsic matrix from depth camera's XYZ positions to the color camera.
            xyz_to_color_camera_extrinsic_matrix (np.ndarray): Extrinsic matrix from depth camera's XYZ positions to the depth camera.
        """
        self.color_frame = color_frame
        self._depth_frame_xyz = depth_frame_xyz
        self.depth_frame = np.sqrt(np.sum(np.square(depth_frame_xyz), axis=-1)).astype(
            np.uint16
        )  # Represent the depth frame as the L2 norm, and convert to mono16

        self.timestamp_millis = timestamp_millis
        self._color_camera_intrinsic_matrix = color_camera_intrinsic_matrix
        self._color_camera_distortion_coeffs = color_camera_distortion_coeffs
        self._depth_camera_intrinsic_matrix = depth_camera_intrinsic_matrix
        self._depth_camera_distortion_coeffs = depth_camera_distortion_coeffs
        self._xyz_to_color_camera_extrinsic_matrix = (
            xyz_to_color_camera_extrinsic_matrix
        )
        self._xyz_to_depth_camera_extrinsic_matrix = (
            xyz_to_depth_camera_extrinsic_matrix
        )
        self._color_to_depth_extrinsic_matrix = (
            xyz_to_depth_camera_extrinsic_matrix
            @ invert_extrinsic_matrix(xyz_to_color_camera_extrinsic_matrix)
        )

    def get_corresponding_depth_pixel_deprecated(
        self, color_pixel: Tuple[int, int]
    ) -> Tuple[int, int]:
        """
        Given an (x, y) coordinate in the color frame, return the corresponding (x, y) coordinate in the depth frame.

        Args:
            color_pixel (Sequence[int]): (x, y) coordinate in the color frame.

        Returns:
            Tuple[int, int]: (x, y) coordinate in the depth frame.
        """

        # Undistort the pixel coordinate in color camera
        distorted_color_pixel = np.array([[color_pixel]], dtype=np.float32)
        undistorted_color_pixel = cv2.undistortPoints(
            distorted_color_pixel,
            self._color_camera_intrinsic_matrix,
            self._color_camera_distortion_coeffs,
            P=self._color_camera_intrinsic_matrix,
        ).reshape(-1)

        # Convert to normalized camera coordinates
        normalized_color_pixel = np.linalg.inv(
            self._color_camera_intrinsic_matrix
        ) @ np.append(undistorted_color_pixel, 1)
        print(f"normalized_color_pixel = {normalized_color_pixel}")

        # Transform the normalized coordinates to depth camera using the extrinsic matrix
        normalized_color_pixel *= DEPTH_MIN_MM
        transformed_homogeneous_depth_pixel = (
            self._color_to_depth_extrinsic_matrix @ np.append(normalized_color_pixel, 1)
        )
        print(
            f"transformed_homogeneous_depth_pixel = {transformed_homogeneous_depth_pixel}"
        )
        normalized_depth_pixel = (
            transformed_homogeneous_depth_pixel[:3]
            / transformed_homogeneous_depth_pixel[3]
        )
        print(f"normalized_depth_pixel = {normalized_depth_pixel}")

        # Convert to pixel coordinates in depth camera
        undistorted_depth_pixel = (
            self._depth_camera_intrinsic_matrix @ normalized_depth_pixel[:3]
        )
        undistorted_depth_pixel = (
            undistorted_depth_pixel[:2] / undistorted_depth_pixel[2]
        )
        print(f"undistorted_depth_pixel = {undistorted_depth_pixel}")

        # Apply distortion to the pixel coordinates in depth camera
        distorted_depth_pixel = distort_pixel_coords(
            undistorted_depth_pixel,
            self._depth_camera_intrinsic_matrix,
            self._depth_camera_distortion_coeffs,
        )
        print(f"distorted_depth_pixel = {distorted_depth_pixel}")

        return (
            round(distorted_depth_pixel[0]),
            round(distorted_depth_pixel[1]),
        )

    def get_corresponding_depth_pixel(
        self, color_pixel: Tuple[int, int]
    ) -> Tuple[int, int]:
        """
        Given an (x, y) coordinate in the color frame, return the corresponding (x, y) coordinate in the depth frame.

        The general approach is as follows:
            1. Deproject the color image pixel coordinate to two positions in the color camera-space: one that corresponds to the
               position at the minimum depth, and one at the maximum depth.
            2. Transform the two positions from color camera-space to depth camera-space.
            3. Project the two positions to their respective depth image pixel coordinates.
            4. The target lies somewhere along the line formed by the two pixel coordinates found in the previous step. We
               iteratively move pixel by pixel along this line. For each depth image pixel, we grab the xyz data at the pixel,
               project it onto the color image plane, and see how far it is from the original color pixel coordinate. We find
               and return the closest match.

        Note that in order to achieve the above, we require two extrinsic matrices - one for projecting the xyz positions to
        the color camera image plane, and one for projecting the xyz positions to the depth camera image plane.

        Args:
            color_pixel (Sequence[int]): (x, y) coordinate in the color frame.

        Returns:
            Tuple[int, int]: (x, y) coordinate in the depth frame.
        """

        def deproject_pixel(pixel, depth, camera_matrix, distortion_coeffs):
            # Normalized, undistorted pixel coord
            pixel = np.array([[[pixel[0], pixel[1]]]], dtype=np.float32)
            pixel_undistorted = cv2.undistortPoints(
                pixel,
                camera_matrix,
                distortion_coeffs,
            )
            return np.array(
                [
                    pixel_undistorted[0][0][0] * depth,
                    pixel_undistorted[0][0][1] * depth,
                    depth,
                ]
            )

        def transform_position(position, extrinsic_matrix):
            position = np.array(position)
            homogeneous_position = np.append(position, 1)
            return np.dot(extrinsic_matrix, homogeneous_position)[:3]

        def project_position(
            position, camera_matrix, distortion_coeffs, extrinsic_matrix=None
        ):
            R = extrinsic_matrix[:3, :3] if extrinsic_matrix is not None else np.eye(3)
            t = extrinsic_matrix[:3, 3] if extrinsic_matrix is not None else np.zeros(3)
            pixels, _ = cv2.projectPoints(
                np.array([[position]]),
                R,
                t,
                camera_matrix,
                distortion_coeffs,
            )
            pixel = pixels[0].flatten()
            return (round(pixel[0]), round(pixel[1]))

        def adjust_pixel_to_bounds(pixel, width, height):
            x = max(0, min(round(pixel[0]), width - 1))
            y = max(0, min(round(pixel[1]), height - 1))
            return (x, y)

        def next_pixel_in_line(curr, start, end):
            # Move one pixel from curr to end
            curr = np.array(curr)
            end = np.array(end)
            direction = end - curr
            direction = direction / np.linalg.norm(direction)
            next = curr + direction
            return (round(next[0]), round(next[1]))

        def is_pixel_in_line(curr, start, end):
            min_x = min(start[0], end[0])
            max_x = max(start[0], end[0])
            min_y = min(start[1], end[1])
            max_y = max(start[1], end[1])

            return min_x <= curr[0] <= max_x and min_y <= curr[1] <= max_y

        # Min-depth and max-depth positions in color camera-space
        min_depth_position_color_space = deproject_pixel(
            color_pixel,
            DEPTH_MIN_MM,
            self._color_camera_intrinsic_matrix,
            self._color_camera_distortion_coeffs,
        )
        max_depth_position_color_space = deproject_pixel(
            color_pixel,
            DEPTH_MAX_MM,
            self._color_camera_intrinsic_matrix,
            self._color_camera_distortion_coeffs,
        )

        # Min-depth and max-depth positions in depth camera-space
        min_depth_position_depth_space = transform_position(
            min_depth_position_color_space, self._color_to_depth_extrinsic_matrix
        )
        max_depth_position_depth_space = transform_position(
            max_depth_position_color_space, self._color_to_depth_extrinsic_matrix
        )

        # Project depth camera-space positions to depth pixels
        min_depth_pixel = project_position(
            min_depth_position_depth_space,
            self._depth_camera_intrinsic_matrix,
            self._depth_camera_distortion_coeffs,
        )
        max_depth_pixel = project_position(
            max_depth_position_depth_space,
            self._depth_camera_intrinsic_matrix,
            self._depth_camera_distortion_coeffs,
        )

        # Make sure pixel coords are in boundary
        depth_frame_height, depth_frame_width, depth_channels = (
            self._depth_frame_xyz.shape
        )
        min_depth_pixel = adjust_pixel_to_bounds(
            min_depth_pixel, depth_frame_width, depth_frame_height
        )
        max_depth_pixel = adjust_pixel_to_bounds(
            max_depth_pixel, depth_frame_width, depth_frame_height
        )

        # Search along the line for the depth pixel for which its corresponding projected color pixel is the closest
        # to the target color pixel
        min_dist = -1
        closest_depth_pixel = min_depth_pixel
        curr_depth_pixel = min_depth_pixel
        while True:
            xyz_mm = self._depth_frame_xyz[curr_depth_pixel[1]][curr_depth_pixel[0]]
            curr_color_pixel = project_position(
                xyz_mm,
                self._color_camera_intrinsic_matrix,
                self._color_camera_distortion_coeffs,
                self._xyz_to_color_camera_extrinsic_matrix,
            )
            distance = np.linalg.norm(
                np.array(curr_color_pixel).astype(float)
                - np.array(color_pixel).astype(float)
            )
            if distance < min_dist or min_dist < 0:
                min_dist = distance
                closest_depth_pixel = curr_depth_pixel

            # Stop if we've processed the max_depth_pixel
            if (
                curr_depth_pixel[0] == max_depth_pixel[0]
                and curr_depth_pixel[1] == max_depth_pixel[1]
            ):
                break

            # Otherwise, find the next pixel along the line we should try
            curr_depth_pixel = next_pixel_in_line(
                curr_depth_pixel, min_depth_pixel, max_depth_pixel
            )
            if not is_pixel_in_line(curr_depth_pixel, min_depth_pixel, max_depth_pixel):
                break

        return closest_depth_pixel

    def get_position(
        self, color_pixel: Tuple[int, int]
    ) -> Optional[Tuple[float, float, float]]:
        """
        Given an (x, y) coordinate in the color frame, return the (x, y, z) position with respect to the camera.

        Args:
            color_pixel (Sequence[int]): (x, y) coordinate in the color frame.

        Returns:
            Optional[Tuple[float, float, float]]: (x, y, z) position with respect to the camera, or None if the position could not be determined.
        """
        depth_pixel = self.get_corresponding_depth_pixel(color_pixel)
        position = self._depth_frame_xyz[depth_pixel[1]][depth_pixel[0]]
        # Negative depth indicates an invalid position
        if position[2] < 0.0:
            return None

        return (float(position[0]), float(position[1]), float(position[2]))

File: ros2/camera_control_cpp/src/camera/lucid_frame.cpp
#include "camera_control_cpp/camera/lucid_frame.hpp"

#include "camera_control_cpp/camera/calibration.hpp"
#include "spdlog/spdlog.h"

namespace {

std::optional<cv::Vec3f> deprojectPixel(const cv::Point2i& pixel, float depth,
                                        const cv::Mat& cameraMatrix,
                                        const cv::Mat& distCoeffs) {
  std::vector<cv::Point2f> src{cv::Point2f(pixel)};
  std::vector<cv::Point2f> undistorted;
  cv::undistortPoints(src, undistorted, cameraMatrix, distCoeffs);

  // cv::undistortPoints can return nan or inf values
  if (undistorted.empty() || std::isnan(undistorted[0].x) ||
      std::isinf(undistorted[0].x) || std::isnan(undistorted[0].y) ||
      std::isinf(undistorted[0].y)) {
    return std::nullopt;
  }

  return cv::Vec3f{undistorted[0].x * depth, undistorted[0].y * depth, depth};
}

cv::Vec3f transformPosition(const cv::Vec3f& position,
                            const cv::Mat& extrinsic) {
  cv::Vec4d homogeneousPosition{position[0], position[1], position[2], 1.0};
  cv::Vec4d transformed;
  for (int i = 0; i < 4; ++i) {
    transformed[i] = 0;
    for (int j = 0; j < 4; ++j) {
      transformed[i] += extrinsic.at<double>(i, j) * homogeneousPosition[j];
    }
  }
  return cv::Vec3f{static_cast<float>(transformed[0]),
                   static_cast<float>(transformed[1]),
                   static_cast<float>(transformed[2])};
}

std::optional<cv::Point2i> projectPosition(
    const cv::Vec3f& position, const cv::Mat& cameraMatrix,
    const cv::Mat& distCoeffs, const cv::Mat& extrinsicMatrix = cv::Mat()) {
  cv::Mat rvec, tvec;
  if (!extrinsicMatrix.empty()) {
    // Extract rotation (3x3) and translation (3x1) from extrinsic
    cv::Mat R = extrinsicMatrix(cv::Range(0, 3), cv::Range(0, 3));
    cv::Mat t = extrinsicMatrix(cv::Range(0, 3), cv::Range(3, 4));

    // Convert rotation matrix to rotation vector
    cv::Rodrigues(R, rvec);
    tvec = t.clone();
  } else {
    rvec = cv::Mat::zeros(3, 1, CV_64F);
    tvec = cv::Mat::zeros(3, 1, CV_64F);
  }

  std::vector<cv::Point3f> objectPoints{
      {cv::Point3f{position[0], position[1], position[2]}}};
  std::vector<cv::Point2f> imagePoints;
  cv::projectPoints(objectPoints, rvec, tvec, cameraMatrix, distCoeffs,
                    imagePoints);

  // cv::projectPoints can return nan or inf values
  if (imagePoints.empty() || std::isnan(imagePoints[0].x) ||
      std::isinf(imagePoints[0].x) || std::isnan(imagePoints[0].y) ||
      std::isinf(imagePoints[0].y)) {
    return std::nullopt;
  }

  return cv::Point2i{static_cast<int>(std::round(imagePoints[0].x)),
                     static_cast<int>(std::round(imagePoints[0].y))};
}

cv::Point2i adjustPixelToBounds(const cv::Point2i& pixel, int width,
                                int height) {
  int x{
      std::max(0, std::min(static_cast<int>(std::round(pixel.x)), width - 1))};
  int y{
      std::max(0, std::min(static_cast<int>(std::round(pixel.y)), height - 1))};
  return cv::Point2i{x, y};
}

cv::Point2i nextPixelInLine(const cv::Point2i& curr, const cv::Point2i& end) {
  cv::Point2f direction{cv::Point2f(end - curr)};
  double norm{cv::norm(direction)};
  if (norm == 0.0) {
    // Already at the end
    return cv::Point2i(static_cast<int>(std::round(curr.x)),
                       static_cast<int>(std::round(curr.y)));
  }
  direction /= norm;
  cv::Point2f next{cv::Point2f(curr) + direction};
  return cv::Point2i(static_cast<int>(std::round(next.x)),
                     static_cast<int>(std::round(next.y)));
}

bool isPixelInLine(const cv::Point2i& curr, const cv::Point2i& start,
                   const cv::Point2i& end) {
  int minX{std::min(start.x, end.x)};
  int maxX{std::max(start.x, end.x)};
  int minY{std::min(start.y, end.y)};
  int maxY{std::max(start.y, end.y)};
  return (minX <= curr.x && curr.x <= maxX) &&
         (minY <= curr.y && curr.y <= maxY);
}

}  // namespace

LucidFrame::LucidFrame(const cv::Mat& colorFrame, const cv::Mat& depthFrameXyz,
                       double timestampMillis,
                       const cv::Mat& colorCameraIntrinsicMatrix,
                       const cv::Mat& colorCameraDistortionCoeffs,
                       const cv::Mat& depthCameraIntrinsicMatrix,
                       const cv::Mat& depthCameraDistortionCoeffs,
                       const cv::Mat& xyzToColorCameraExtrinsicMatrix,
                       const cv::Mat& xyzToDepthCameraExtrinsicMatrix,
                       std::pair<int, int> colorFrameOffset)
    : colorFrame_(colorFrame),
      depthFrameXyz_(depthFrameXyz),
      timestampMillis_(timestampMillis),
      colorCameraIntrinsicMatrix_(colorCameraIntrinsicMatrix),
      colorCameraDistortionCoeffs_(colorCameraDistortionCoeffs),
      depthCameraIntrinsicMatrix_(depthCameraIntrinsicMatrix),
      depthCameraDistortionCoeffs_(depthCameraDistortionCoeffs),
      xyzToColorCameraExtrinsicMatrix_(xyzToColorCameraExtrinsicMatrix),
      xyzToDepthCameraExtrinsicMatrix_(xyzToDepthCameraExtrinsicMatrix),
      colorFrameOffset_(colorFrameOffset) {
  colorToDepthExtrinsicMatrix_ =
      xyzToDepthCameraExtrinsicMatrix_ *
      calibration::invertExtrinsicMatrix(xyzToColorCameraExtrinsicMatrix_);
}

LucidFrame::~LucidFrame() {}

cv::Mat LucidFrame::getDepthFrame() const {
  // Split into separate X, Y, Z components
  std::vector<cv::Mat> channels(3);
  cv::split(depthFrameXyz_, channels);

  // Compute the L2 norm
  cv::Mat depthFrame;
  depthFrame = channels[0].mul(channels[0]) + channels[1].mul(channels[1]) +
               channels[2].mul(channels[2]);
  cv::sqrt(depthFrame, depthFrame);

  // Convert to 16-bit unsigned integer (mono16 format)
  cv::Mat depthFrameMono16;
  depthFrame.convertTo(depthFrameMono16, CV_16U);

  return depthFrameMono16;
}

std::optional<cv::Point2i> LucidFrame::getCorrespondingDepthPixel(
    const cv::Point2i& colorPixel) const {
  /*
  Given an (x, y) coordinate in the color frame, return the corresponding (x, y)
  coordinate in the depth frame.

  The general approach is as follows:
      1. Deproject the color image pixel coordinate to two positions in the
  color camera-space: one that corresponds to the position at the minimum depth,
  and one at the maximum depth.
      2. Transform the two positions from color camera-space to depth
  camera-space.
      3. Project the two positions to their respective depth image pixel
  coordinates.
      4. The target lies somewhere along the line formed by the two pixel
  coordinates found in the previous step. We iteratively move pixel by pixel
  along this line. For each depth image pixel, we grab the xyz data at the
  pixel, project it onto the color image plane, and see how far it is from the
  original color pixel coordinate. We find and return the closest match.
  Note that in order to achieve the above, we require two extrinsic matrices -
  one for projecting the xyz positions to the color camera image plane, and one
  for projecting the xyz positions to the depth camera image plane.
  */

  // Apply frame ROI offset. The calibration matrices and coefficients were
  // calculated based on the max camera frame size, and if there was a reduced
  // ROI set when capturing the frame, the pixel coordinate for the frame must
  // be converted to that of the full size frame.
  cv::Point2i fullFrameColorPixel{
      colorPixel +
      cv::Point2i(colorFrameOffset_.first, colorFrameOffset_.second)};

  // Min-depth and max-depth positions in color camera-space
  auto minDepthPositionColorSpaceOpt{deprojectPixel(
      fullFrameColorPixel, DEPTH_MIN_MM, colorCameraIntrinsicMatrix_,
      colorCameraDistortionCoeffs_)};
  if (!minDepthPositionColorSpaceOpt) {
    return std::nullopt;
  }
  cv::Vec3f minDepthPositionColorSpace{minDepthPositionColorSpaceOpt.value()};
  auto maxDepthPositionColorSpaceOpt{deprojectPixel(
      fullFrameColorPixel, DEPTH_MAX_MM, colorCameraIntrinsicMatrix_,
      colorCameraDistortionCoeffs_)};
  if (!maxDepthPositionColorSpaceOpt) {
    return std::nullopt;
  }
  cv::Vec3f maxDepthPositionColorSpace{maxDepthPositionColorSpaceOpt.value()};

  // Min-depth and max-depth positions in depth camera-space
  cv::Vec3f minDepthPositionDepthSpace{transformPosition(
      minDepthPositionColorSpace, colorToDepthExtrinsicMatrix_)};
  cv::Vec3f maxDepthPositionDepthSpace{transformPosition(
      maxDepthPositionColorSpace, colorToDepthExtrinsicMatrix_)};

  // Project depth camera-space positions to depth pixels
  auto minDepthPixelOpt{projectPosition(minDepthPositionDepthSpace,
                                        depthCameraIntrinsicMatrix_,
                                        depthCameraDistortionCoeffs_)};
  if (!minDepthPixelOpt) {
    return std::nullopt;
  }
  cv::Point2i minDepthPixel{minDepthPixelOpt.value()};
  auto maxDepthPixelOpt{projectPosition(maxDepthPositionDepthSpace,
                                        depthCameraIntrinsicMatrix_,
                                        depthCameraDistortionCoeffs_)};
  if (!maxDepthPixelOpt) {
    return std::nullopt;
  }
  cv::Point2i maxDepthPixel{maxDepthPixelOpt.value()};

  // Make sure pixel coords are in boundary
  cv::Point2i minDepthPixelAdjusted{adjustPixelToBounds(
      minDepthPixel, depthFrameXyz_.cols, depthFrameXyz_.rows)};
  cv::Point2i maxDepthPixelAdjusted{adjustPixelToBounds(
      maxDepthPixel, depthFrameXyz_.cols, depthFrameXyz_.rows)};

  // Search along the line for the depth pixel for which its corresponding
  // projected color pixel is the closest to the target color pixel
  int minDist{-1};
  cv::Point2i closestDepthPixel{minDepthPixelAdjusted};
  cv::Point2i currDepthPixel{minDepthPixelAdjusted};
  while (true) {
    cv::Vec3f xyzmm{
        depthFrameXyz_.at<cv::Vec3f>(currDepthPixel.y, currDepthPixel.x)};
    auto currColorPixelOpt{projectPosition(xyzmm, colorCameraIntrinsicMatrix_,
                                           colorCameraDistortionCoeffs_,
                                           xyzToColorCameraExtrinsicMatrix_)};
    if (!currColorPixelOpt) {
      return std::nullopt;
    }
    cv::Point2i currColorPixel{currColorPixelOpt.value()};
    double distance{cv::norm(cv::Point2f(currColorPixel) -
                             cv::Point2f(fullFrameColorPixel))};
    if (distance < minDist || minDist < 0) {
      minDist = distance;
      closestDepthPixel = currDepthPixel;
    }
    // Stop if we've processed the maxDepthPixel
    if (currDepthPixel.x == maxDepthPixel.x &&
        currDepthPixel.y == maxDepthPixel.y) {
      break;
    }
    // Otherwise, find the next pixel along the line that we shold try
    currDepthPixel = nextPixelInLine(currDepthPixel, maxDepthPixel);
    if (!isPixelInLine(currDepthPixel, minDepthPixel, maxDepthPixel)) {
      break;
    }
  }

  return closestDepthPixel;
}

std::optional<cv::Vec3f> LucidFrame::getPosition(
    const cv::Point2i& colorPixel) const {
  auto depthPixelOpt{getCorrespondingDepthPixel(colorPixel)};
  if (!depthPixelOpt) {
    return std::nullopt;
  }
  cv::Point2i depthPixel{depthPixelOpt.value()};

  cv::Vec3f position{depthFrameXyz_.at<cv::Vec3f>(depthPixel.y, depthPixel.x)};
  // Negative depth indicates an invalid position. Depth greater than 2^14 - 1
  // also indicates invalid position.
  if (position[2] < 0.0f || position[2] > ((1 << 14) - 1)) {
    spdlog::warn("Invalid depth value at pixel ({}, {}): {}", depthPixel.x,
                 depthPixel.y, position[2]);
    return std::nullopt;
  }

  return position;
}
